providers/provider_set.go

```diff
@@ -10,25 +10,30 @@ import (
 // It is used as an intermediary data struct between what is stored in the datastore
 // and the list of providers that get passed to the consumer of a .GetProviders call
 type providerSet struct {
-	providers []peer.ID
-	set       map[peer.ID]time.Time
+	providers      []peer.ID
+	set            map[peer.ID]time.Time
+	keyToProviders map[string][]peer.ID // TODO change to map of maps, otherwise there can be duplicate peer IDs
```

Comment:
actually now that I thought about it, I think these are okay as slices because there can't be duplicate provider records in the datastore, since the DB key is the provider ID + provided CID essentially. so when these slices are written to while the database is being iterated over, it shouldn't be possible to have duplicate peer IDs.